src/NUnitEngine/nunit.engine.tests/Runners/MasterTestRunnerTests.cs

```diff
@@ -170,40 +263,106 @@ public void RunAsync()
-                Assert.That(suite.GetAttribute("failed", 0), Is.EqualTo(MockAssembly.Failed));
-                Assert.That(suite.GetAttribute("skipped", 0), Is.EqualTo(MockAssembly.Skipped));
-                Assert.That(suite.GetAttribute("inconclusive", 0), Is.EqualTo(MockAssembly.Inconclusive));
+                Assert.That(suite.GetAttribute("passed", 0), Is.EqualTo(data.Passed));
+                Assert.That(suite.GetAttribute("failed", 0), Is.EqualTo(data.Failed));
+                Assert.That(suite.GetAttribute("skipped", 0), Is.EqualTo(data.Skipped));
+                Assert.That(suite.GetAttribute("inconclusive", 0), Is.EqualTo(data.Inconclusive));
             }
 
+            AssertThatIdsAreUnique(result);
+
             Assert.That(_events[0].Name, Is.EqualTo("start-run"));
-            Assert.That(_events[0].GetAttribute("count", -1), Is.EqualTo(MockAssembly.Tests * _numAssemblies), "Start-run count value");
-            Assert.That(_events[1].Name, Is.EqualTo("start-suite"));
-            Assert.That(_events[_events.Count - 2].Name, Is.EqualTo("test-suite"));
+            Assert.That(_events[0].GetAttribute("count", -1), Is.EqualTo(_totalTests), "Start-run count value");
             Assert.That(_events[_events.Count - 1].Name, Is.EqualTo("test-run"));
-            Assert.That(_events.Count(x => x.Name == "test-case"), Is.EqualTo(MockAssembly.Tests * _numAssemblies));
+            Assert.That(_events.Count(x => x.Name == "test-case"), Is.EqualTo(_totalTests));
+        }
+
+        private void AssertThatIdsAreUnique(XmlNode test)
+        {
+            AssertThatIdsAreUnique(test, new Dictionary<string, bool>());
+        }
+
+        private void AssertThatIdsAreUnique(XmlNode test, Dictionary<string, bool> dict)
+        {
+            Assert.That(dict, Does.Not.ContainKey(test.GetAttribute("id")));
+
+            foreach (XmlNode child in test.SelectNodes("test-suite"))
+                AssertThatIdsAreUnique(child, dict);
         }
 
         void ITestEventListener.OnTestEvent(string report)
         {
             _events.Add(XmlHelper.CreateXmlNode(report));
         }
+
+        #endregion
+
+        #region Nested Helper Classes
+
+        public class TestRunData
+        {
+            public string CommandLine;
+            public Type ExpectedRunner;
+            public AssemblyData[] AssemblyData;
+
+            public TestRunData(string commandLine, Type expectedRunner, params AssemblyData[] assemblyData)
+            {
+                CommandLine = commandLine;
+                ExpectedRunner = expectedRunner;
+                AssemblyData = assemblyData;
+            }
+
+            public override string ToString()
+            {
+                var sb = new StringBuilder(ExpectedRunner.Name);
+                foreach (var assembly in AssemblyData)
+                    sb.Append($", {assembly.Name}");
+                return sb.ToString();
```

Comment:
I like this `TestRunData` class, compared to the first iteration of these tests in the commits. Good thinking adding the `ToString()` as well! 😄